frontend/ui/menusorter.lua

```diff
@@ -0,0 +1,166 @@
+
+        if lfs.attributes(menu_order..".lua") then
+            return require(menu_order) or {}
+        end
+    end
+    return {}
+end
+
+function MenuSorter:mergeAndSort(config_prefix, item_table, order)
+    local user_order = self:readMSSettings(config_prefix)
+    if user_order then
+        for user_order_id,user_order_item in pairs(user_order) do
+            if order[user_order_id] then
+                order[user_order_id] = user_order_item
+            end
+        end
+    end
+    return self:sort(item_table, order)
+end
+
+function MenuSorter:sort(item_table, order)
+    local menu_table = {}
+    local sub_menus = {}
+    -- the actual sorting of menu items
+    for order_id, order_item in pairs (order) do
+        -- user might define non-existing menu item
+        if item_table[order_id] ~= nil then
+            local tmp_menu_table = {}
+            menu_table[order_id] = item_table[order_id]
+            menu_table[order_id].id = order_id
+            for order_number,order_number_id in ipairs(order_item) do
+                -- this is a submenu, mark it for later
+                if order[order_number_id] then
+                    table.insert(sub_menus, order_number_id)
+                    tmp_menu_table[order_number] = {
+                        id = order_number_id,
+                    }
+                -- regular, just insert a menu action
+                else
+                    if order_number_id == "----------------------------" then
+                        -- it's a separator
+                        tmp_menu_table[order_number] = self.separator
+                    elseif item_table[order_number_id] ~= nil then
+                        item_table[order_number_id].id = order_number_id
+                        tmp_menu_table[order_number] = item_table[order_number_id]
+                        -- remove reference from item_table so it won't show up as orphaned
+                        item_table[order_number_id] = nil
+                    end
+                end
+            end
+            -- compress menus
+            -- if menu_items were missing we might have a table with gaps
+            -- but ipairs doesn't like that and quits when it hits nil
+            local i = 1
+            local new_index = 1
+            while i <= table.maxn(tmp_menu_table) do
+                local v = tmp_menu_table[i]
+                if v then
+                    if v.id == "----------------------------" then
+                        new_index = new_index - 1
+                        menu_table[order_id][new_index].separator = true
+                    else
+                        -- fix the index
+                        menu_table[order_id][new_index] = tmp_menu_table[i]
+                    end
+
+                    new_index = new_index + 1
+                end
+                i = i + 1
+            end
+        else
+            DEBUG("menu id not found:", order_id)
```

Comment:
DEBUG is deprecated, use logger instead. Missing id is probably `logger.warn` worthy ;)